pkg/controllers/util/federatedclient/podinformer.go

```diff
@@ -0,0 +1,109 @@
+package federatedclient
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (f *federatedClientFactory) addPodInformer(informer informers.SharedInformerFactory, client kubeclient.Interface) {
+	informer.InformerFor(&corev1.Pod{}, func(k kubeclient.Interface, duration time.Duration) cache.SharedIndexInformer {
+		return cache.NewSharedIndexInformer(
+			f.podListerWatcher(client),
+			&corev1.Pod{},
+			0,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		)
+	})
+}
+
+func (f *federatedClientFactory) podListerWatcher(client kubeclient.Interface) cache.ListerWatcher {
+	return &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			cur := atomic.LoadInt32(&f.availablePodListers)
+			if cur == 0 {
+				return nil, fmt.Errorf("do not have enought lister")
```

Comment:
What happens when this returns error? Does `Reflector` backoff-retry? Is it better to simply use use a `make(chan struct{}, maxPodListers)` as the semaphore?

Reply:
- What happens when this returns error? Does `Reflector` backoff-retry?
Just logs an error. It will retry, but not sure backoff or just poll. You can see that in test case `4 clusters, limit 2`.

- Is it better to simply use use a `make(chan struct{}, maxPodListers)` as the semaphore?
Do you mean block it when there is not enough lister? Like this:
```go
ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
        ch <- struct{}{}
        defer <-ch
        // do list here...
}
```

Reply:
Agree with @SOF3 here, we should manage the concurrency here explicitly instead of relying on `Reflector` backoff retry logic, which may not be suited for a use case like this. I think the channel method @JackZxj suggested would work; alternatively we can use `sync.Semaphore`.

Reply:
+1 to blocking with a semaphore